package umpaz.brewinandchewin.platform;

import com.mojang.datafixers.util.Pair;
import com.mojang.serialization.Codec;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.core.HolderLookup;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.chat.Component;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.Identifier;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.MenuProvider;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.food.FoodProperties;
import net.minecraft.world.inventory.InventoryMenu;
import net.minecraft.world.inventory.MenuType;
import net.minecraft.world.inventory.Slot;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.crafting.Ingredient;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.material.Fluid;
import org.jetbrains.annotations.Nullable;
import umpaz.brewinandchewin.common.attachment.RagingAttachment;
import umpaz.brewinandchewin.common.attachment.TipsyHeartsAttachment;
import umpaz.brewinandchewin.common.block.entity.AgingCaskBlockEntity;
import umpaz.brewinandchewin.common.block.entity.KegBlockEntity;
import umpaz.brewinandchewin.common.container.AbstractedFluidTank;
import umpaz.brewinandchewin.common.container.AbstractedItemHandler;
import umpaz.brewinandchewin.common.block.entity.container.AgingCaskMenu;
import umpaz.brewinandchewin.common.block.entity.container.KegMenu;
import umpaz.brewinandchewin.common.block.entity.container.SidedKegWrapper;
import umpaz.brewinandchewin.common.utility.BnCMenuConstructor;
import umpaz.brewinandchewin.common.utility.AbstractedFluidIngredient;
import umpaz.brewinandchewin.common.utility.AbstractedFluidStack;
import umpaz.brewinandchewin.common.utility.KegRecipeWrapper;

import java.util.List;
import java.util.function.BiConsumer;

public interface BnCPlatformHelper {

    BnCPlatform getPlatform();

    boolean isModLoaded(String modId);

    default boolean isModLoadedEarly(String modId) {
        return isModLoaded(modId);
    }

    boolean isDevelopmentEnvironment();

    void sendClientbound(ServerPlayer player, CustomPacketPayload payload);
    void sendClientboundTracking(Entity tracked, CustomPacketPayload payload);
    void sendServerbound(CustomPacketPayload payload);

    Component getFluidDisplayName(AbstractedFluidStack wrapper);

    void openKegMenu(Player player, KegBlockEntity blockEntity, BlockPos pos);

    void openBlockPosMenu(Player player, MenuProvider menuProvider, BlockPos pos);

    BlockEntityType.BlockEntitySupplier<KegBlockEntity> supplyBlockEntity();
    BlockEntityType.BlockEntitySupplier<AgingCaskBlockEntity> supplyAgingCaskBlockEntity();
    MenuType<KegMenu> createMenuType(BnCMenuConstructor<KegMenu> constructor);
    MenuType<AgingCaskMenu> createAgingCaskMenuType(BnCMenuConstructor<AgingCaskMenu> constructor);

    AbstractedItemHandler createKegInventory(int size, BiConsumer<AbstractedItemHandler, Integer> onContentsChanged);

    AbstractedFluidTank createKegTank(long capacity, Runnable onContentsChanged);

    default Slot createKegSlot(AbstractedItemHandler inventory, int slot, int x, int y) {
        return createKegSlot(inventory, slot, x, y, true, null);
    }
    default Slot createKegContainerSlot(AbstractedItemHandler inventory, int slot, int x, int y) {
        return createKegSlot(inventory, slot, x, y, true, KegMenu.EMPTY_CONTAINER_SLOT_TANKARD);
    }
    default Slot createKegResultSlot(AbstractedItemHandler inventory, int slot, int x, int y) {
        return createKegSlot(inventory, slot, x, y, false, null);
    }
    Slot createKegSlot(AbstractedItemHandler inventory, int slot, int x, int y, boolean canInsert, @Nullable Identifier noItemIcon);


    KegRecipeWrapper createRecipeWrapper(AbstractedItemHandler itemHandler, AbstractedFluidTank fluidTank);

    SidedKegWrapper createSidedKegWrapper(AbstractedItemHandler inventory, Direction direction);

    Codec<AbstractedFluidStack> getFluidStackWrapperCodec();

    StreamCodec<RegistryFriendlyByteBuf, AbstractedFluidStack> getFluidStackWrapperStreamCodec();

    Codec<AbstractedFluidIngredient> getFluidIngredientWrapperCodec();

    StreamCodec<RegistryFriendlyByteBuf, AbstractedFluidIngredient> getFluidIngredientWrapperStreamCodec();

    AbstractedFluidStack deserializeTankFluidStack(CompoundTag tag, HolderLookup.Provider provider);

    ItemStack getCraftingRemainingItem(ItemStack stack);

    void initFluids();
    void initCreativeTab();

    boolean isEdible(ItemStack stack, LivingEntity entity);

    FoodProperties getFoodProperties(ItemStack stack, LivingEntity entity);

    MinecraftServer getServer();

    RagingAttachment getRagingAttachment(LivingEntity entity);
    void setRagingAttachment(LivingEntity entity, @Nullable RagingAttachment value);

    TipsyHeartsAttachment getTipsyHeartsAttachment(LivingEntity entity);
    void setTipsyHeartsAttachment(LivingEntity entity, @Nullable TipsyHeartsAttachment value);

    Object createLoaderFluidStack(AbstractedFluidStack abstracted);
    Object copyLoaderFluidStack(Object fluidStack);

    @Nullable
    AbstractedFluidTank getFluidContainerFromItem(ItemStack stack);

    String getAttachmentKey();

    Fluid getMilkFluid();
    Fluid getFlowingMilkFluid();


    boolean hasFoodEffectTooltip();
}